### PR TITLE
 Updated to React 16.3. Use rect-testing-library instead of enzyme.

### PR DESCRIPTION
**Pull request by ericsoderberghp**
#### What does this PR do?

Changes many components to support using `ref={}`. This was done by taking advantage of `React.forwardRef()`, which is only available in React 16.3.
These changes also necessitated moving away from `enzyme` for testing since it cannot handle `forwardRef` and nobody appears to be supporting enzyme.
Also added storybooks for the components changed.

#### Where should the reviewer start?

TextInput.js

#### What testing has been done on this PR?

unit tests

#### How should this be manually tested?

storybook

#### Do the grommet docs need to be updated?

no?

#### Should this PR be mentioned in the release notes?

definitely

#### Is this change backwards compatible or is it a breaking change?

It requires a peer dependency version of 16.3 for `react` and `react-dom`.
Other than that, there should be no compatibility issues.

### CODE REVIEW

File: src/js/components/Button/button.stories.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      </Grommet>
+    );
+  }
+}
+
+class FocusedButton extends Component {
+  ref = React.createRef()
+
+  componentDidMount() {
+    setTimeout(() => { this.ref.current.focus(); }, 0);
```
Comment (alansouzati) [inline src/js/components/Button/button.stories.js]: why for Anchor story we did not use setTimeout and here we are using setTimeout ?

File: src/js/components/RangeSelector/RangeSelector.js [hunk 2]
```diff
@@ -183,6 +183,9 @@ if (process.env.NODE_ENV !== 'production') {
   doc(RangeSelector);
 }
 
-export default compose(
+const WrappedRangeSelector = compose(
   withTheme,
 )(RangeSelector);
+
+export default React.forwardRef((props, ref) =>
```
Comment (alansouzati) [inline src/js/components/RangeSelector/RangeSelector.js]: how about we do a HOC called `withForwardRef` and we do this once inside that hoc? with that we can keep using compose + export default.

File: src/js/components/hocs.js [hunk 3]
```diff
@@ -10,36 +9,25 @@ export const withFocus = (WrappedComponent) => {
...
       this.state = {
         mouseActive: false,
         focus: false,
       };
     }
-    componentDidMount() {
-      window.addEventListener('mousedown', this.handleActiveMouse);
 
-      // we could be using onFocus in the wrapper node itself
-      // but react does not invoke it if you programically
```
Comment (alansouzati) [inline src/js/components/hocs.js]: where did we move this to?

File: src/js/utils/DOM.js [hunk 4]
```diff
@@ -73,7 +73,7 @@ export const setFocusWithoutScroll = (element) => {
   const x = window.scrollX;
   const y = window.scrollY;
   element.focus();
-  window.scrollTo(x, y);
+  window.scrollTo(x, y); // this doesn't work in unit tests :(
```
Comment (alansouzati) [inline src/js/utils/DOM.js]: can we mock ?

File: src/js/components/hocs.js (hunk 3 above)
Comment (alansouzati) [reply src/js/components/hocs.js]: I just validated the focus stopped working

File: src/js/components/Button/button.stories.js (hunk 1 above)
Comment (ericsoderberghp) [reply src/js/components/Button/button.stories.js]: Good catch. We don't need the timeout. I think this was a copy/paste issue.

File: src/js/components/DropButton/DropButton.js [hunk 5]
```diff
@@ -13,25 +13,37 @@ class DropButton extends Component {
...
+    const { buttonRef, show } = prevState;
+    const nextButtonRef = dropButtonRef || buttonRef;
+    const reRenderOnMount = (show === undefined && open);
+    if (open !== undefined && open !== show) {
+      return { show: open, reRenderOnMount, buttonRef: nextButtonRef };
+    }
+    if (nextButtonRef !== buttonRef) {
+      return { buttonRef: nextButtonRef };
+    }
+    return null;
```
[the remark this answers is not in the document]
Comment (ericsoderberghp) [reply src/js/components/DropButton/DropButton.js]: Can you clarify? Are you suggesting that we test for undefined, that we set the state variable to undefined, or asking whether the user can pass undefined for the `dropButtonRef` property?

File: src/js/components/hocs.js (hunk 3 above)
Comment (ericsoderberghp) [reply src/js/components/hocs.js]: What is a scenario I can use to reproduce the issue?

Comment (ericsoderberghp) [reply src/js/components/hocs.js]: I figured it out. I've restored the behavior.

File: src/js/components/DropButton/DropButton.js [hunk 6]
```diff
@@ -13,25 +13,37 @@ class DropButton extends Component {
...
     }
   }
 
-  componentWillReceiveProps({ open }) {
-    const { show } = this.state;
-    if (open !== undefined && open !== show) {
-      this.setState({ show: open });
+  componentDidUpdate(prevProps, prevState) {
+    if (!this.state.show && prevState.show) {
+      // focus on the button if the drop is closed
```
[the remark this answers is not in the document]
Comment (ericsoderberghp) [reply src/js/components/DropButton/DropButton.js]: It seemed cleaner to me to track this via DropButton. And, I did change it to use `setFocusWithoutScroll`.

**Pull request merged**